handle empty response

garasubo opened the pull request:
When the server sends RST packet without any contents, the incoming
thread will be shutdown with IO error. But outbound thread keeps alive
because it doesn't recognize incoming thread termination.

In this commit, I added channels to notify incoming thread termination
to outbound thread and outbound thread termination to incoming thread.

legokichi commented on src/relay.rs:
```diff
@@ -12,10 +12,6 @@ use crate::thread::spawn_thread;
-    /// client address
-    client_addr: SocketAddr,
-    /// server address
-    server_addr: SocketAddr,
```
unused fields

garasubo commented on src/relay.rs:
```diff
@@ -121,7 +130,16 @@ fn spawn_relay_half(
             Ok(size) => trace!("{}: {} ==> {}: {} bytes", name, src_addr, dst_addr, size),
-            Err(err) if err.kind() == K::WouldBlock || err.kind() == K::TimedOut => {}
+            Err(err) if err.kind() == K::WouldBlock || err.kind() == K::TimedOut => {
+                match another_rx.try_recv() {
```
Sounds a good idea. Fixed

eldesh reviewed (changes requested):
The shutdown_relay_by_connection_rest passes on the master branch before the merge.
It seems that the fixes are not being tested properly.

garasubo commented:
@eldesh It seems all tests passed in this branch as well.
https://app.circleci.com/pipelines/github/Idein/gatekeeper/205/workflows/57a113cb-79b7-4022-a4ad-15831adae9d1
Is there any tests not running in this CI?

legokichi reviewed (dismissed):
dante-server seems to pass RSTs transparently.
However, I think this PR is fine as long as the connection can be simply disconnected

eldesh commented:
> It seems all tests passed in this branch as well.

Yes I see. It has passed all test cases.

But doesn't this PR fix bugs in the previous version?
If so, then the newly added test (`shutdown_relay_by_connection_rest`) should fail in the **master branch** (i.e, in the pre-fixed version), right?

garasubo commented:
I see. Added Node.js based test. I firstly tried to add unit test in Rust, but it is a little bit difficult since there is no straight way to close TCP session with RST packet using std. So, I added a snippet, which was used to find this bug.

eldesh commented on example.yml:
```diff
@@ -55,3 +55,25 @@
     port: Any
     protocol:
       Specif: Tcp
+- Allow:
```
Is this change as intended?

garasubo replied:
No, removed

eldesh reviewed (dismissed):
I have confirmed that the test fails in previous versions.
Looks good :+1:
